### ISSUE TITLE
getpwnam: No such file or directory

### ISSUE THREAD

**Issue by dradtke** [thestinger/playpen#47]
I'm trying to run playpen on openSUSE 13.1, but I keep running into some issues. Building it was fairly easy (though it did require a small update to the Makefile), but whenever I run it I get this error message:

```
playpen: getpwnam: No such file or directory
playpen: application terminated with error code 1
```

Debugging this has been tricky as well, since when I try to get a backtrace, GDB complains about an import error (which I'm pretty sure is related to [this bug](https://bugzilla.novell.com/show_bug.cgi?id=857541)). I do have some [strace output](http://pastebin.com/QXGrpCFX), but nothing there seems very useful.

Any help would be appreciated!

**Comment by thestinger** [comment]
It expects a `passwd` database in the chroot so it can switch to a non-root user (`nobody` by default). It seems like it's missing in that root directory.

**Comment by dradtke** [comment]
So is that just a matter of creating an /etc/passwd file relative to the sandbox?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   From: Daniel MicaySent: Sunday, February 22, 2015 12:18 AMTo: thestinger/playpenReply To: thestinger/playpenCc: Damien RadtkeSubject: Re: [playpen] getpwnam: No such file or directory (#47)It expects a passwd database in the chroot so it can switch to a non-root user (nobody by default). It seems like it's missing in that root directory.

—Reply to this email directly or view it on GitHub.

**Comment by thestinger** [comment]
Yeah it just needs enough for `getpwnam` to work (probably `passwd`, `group` and `shadow`).

**Comment by dradtke** [comment]
Creating {{sandbox/etc/passwd}}, {{sandbox/etc/group}}, and {{sandbox/etc/shadow}} from the system copies didn't seem to work.

**Comment by nanthiel** [comment]
You need at least `sandbox/etc/nsswitch.conf` aswell.

This tree works for me (from issue #44):
```
sandbox
├── bin
├── dev
│   └── shm
├── etc
│   ├── java-7-openjdk
│   │   └── amd64
│   │       └── jvm.cfg
│   ├── nsswitch.conf
│   └── passwd
├── extra
├── home
│   └── matej
├── lib
├── lib64
├── proc
├── run
├── sbin
├── srv
├── sys
├── tmp
├── usr
└── var
    └── tmp
```
You probably won't need the Java stuff.

**Comment by dradtke** [comment]
Cool, copying `/etc/nsswitch.conf` over improved things a little bit. Now I'm getting a "no passwd entry for username nobody" error, even though there is a line in `sandbox/etc/passwd` for it. I believe I already had a "nobody" user set up for running PHP or something, so I'm surprised that it's not finding it there. The relevant line looks like this:

```
nobody:x:65534:65533:nobody:/var/lib/nobody:/bin/bash
```

which seems correct to me. Could it somehow be a permissions issue?

**Comment by nanthiel** [comment]
Does a minimal test case work for you? This works for me without problems:

Directory tree:
```
sandbox/
├── bin
├── etc
│   ├── nsswitch.conf
│   └── passwd
├── lib
├── lib64
├── proc
├── tmp
└── usr
```

nsswitch.conf:
```
# Begin /etc/nsswitch.conf

passwd: files
group: files
shadow: files

publickey: files

hosts: files dns myhostname
networks: files

protocols: files
services: files
ethers: files
rpc: files

netgroup: files

# End /etc/nsswitch.conf
```

passwd:
```
nobody:x:99:99:nobody:/:/bin/nologin
```

Command:
```
sudo playpen -p -b /bin -b /lib -b /lib64 -b /usr -S syscalls_all.txt -u nobody sandbox -- echo 123
```

**Comment by dradtke** [comment]
Okay, I got that example to work; had to make a couple changes to `nsswitch.conf` (I think I needed to set passwd to files instead of compat), and it looks like I need to bind `/bin`, `/usr`, and `/lib64` at a minimum.

The title of this issue isn't really relevant anymore, but I am running into one more thing that I can't figure out. =) I'm trying to use the sandbox to run `rustc`; I've copied all the libs and binaries into the sandbox; but it keeps failing with `seccomp_syscall_resolve_num_arch`. So far I've tracked this down to `ptrace()` returning a value of -1, but not setting `errno`, so it tries to pass -1 to seccomp, which then fails. Any idea what could cause `ptrace()` to fail but not report an error?

**Comment by sstewartgallus** [comment]
>   On error, all requests return -1, and errno is  set  appropriately.   Since
       the  value  returned  by  a  successful PTRACE_PEEK* request may be -1, the
       caller must clear errno before the call, and then  check  it  afterward  to
       determine whether or not an error occurred.

Does `seccomp_syscall_resolve_num_arch` use PEEK requrests?

**Comment by dradtke** [comment]
`ptrace` does, and it is properly clearing out `errno` beforehand, so I'm not sure what it means for a "successful" return of -1.

**Comment by thestinger** [comment]
`seccomp_syscall_resolve_num_arch` just resolves a system call number to a name string. It doesn't make a system call.

**Comment by dradtke** [comment]
But -1 isn't a valid system call number, is it? The call to `seccomp_syscall_resolve_num_arch` gets made because `ptrace` doesn't set `errno` to anything, but it still returned -1, which is supposed to indicate "an error occurred, please check errno to see what it was." Hence, the call fails because it's trying to resolve system call -1 to a name string.

It might be as simple a fix as "don't call `seccomp_syscall_resolve_num_arch` with a value of -1, even if `errno` is 0", but it would still be nice to know why `ptrace` is behaving this way.

**Comment by thestinger** [comment]
It's not a valid system call number but so the application might really be passing -1 as the system call number.

**Comment by nanthiel** [comment]
Yes, the contents of the register probably really are `-1`. But could the process get trapped by something other than a syscall? Then the register might not contain a syscall at all.

**Comment by dradtke** [comment]
I think I'm making some progress by ensuring that the returned `name` is non-NULL before attempting to call `seccomp_syscall_resolve_num_arch` (I tried just the case of a -1 syscall, but then it failed on a different syscall number, 318), but now it either seems to hang or cause a crash in the program being called.

**Comment by thestinger** [comment]
`seccomp_syscall_resolve_num_arch` is what returns `name` so I don't know what you mean.

> but then it failed on a different syscall number, 318

You're running a program that's using `getrandom` but your libseccomp doesn't have it in the database yet. You need to use it from Git because they haven't done a release for a while.

**Comment by thestinger** [comment]
@nanthiel: It likely needs to check that `PTRACE_O_TRACESECCOMP` is the reason SIGTRAP happened to handle signals properly. It should detect a signal and then raise with in the `PTRACE_CONT` call. The learning mode is just there for convenience and isn't a very robust feature right now.

**Comment by thestinger** [comment]
I think this has been resolved.

The issue with signals during learning mode has been fixed by 0d7727046ec63b2d6963fb07899f63845fbbfecd but I think the issue here was just an out-of-date libseccomp. The current stable release of the library now has `getrandom`.
